% \VignetteIndexEntry{rPlant}


% This vignette requires validation in order to run.  To compile it in Sweave, you need to enter your credentials as part of this document and then compile.  


\documentclass[12pt]{article}
\usepackage{graphicx}
\usepackage{amsfonts}
\usepackage{enumitem}
\usepackage{amssymb}
\usepackage{hyperref}
\usepackage[fleqn]{amsmath}
\usepackage{natbib}
\usepackage{url}

\bibpunct{(}{)}{;}{a}{}{,}

\setlength{\topmargin}{-1.0 in}
\setlength{\textheight}{9.9  in}
\setlength{\oddsidemargin}{-.2in} 
\setlength{\evensidemargin}{-.2in}
\setlength{\textwidth}{6.5in}

\title{rPlant}
\author{Barb Banbury, University of Tennessee, darwinthesun@gmail.com \\ 
Kurt Michels, University of Arizona, kamichels@math.arizona.edu}
\date{\today}
\def\fs{\footnotesize}

\usepackage{Sweave}
\begin{document}
\maketitle

\tableofcontents

\section{Introduction}

\noindent The iPlant Collaborative has developed many resources to deal with the emerging computational challenges facing biology. The project was initially designed to support the plant sciences, but thanks to a generic approach, can be equally used by other disciplines.  Users have access to many different applications for data analysis, including clustering/network analyses, QTL mapping, sequence alignments, phylogenetic tree building, and comparative methods. \\

\noindent The main interface is its user-friendly Discovery Environment (\url{http://www.iplantcollaborative.org/discover/discovery-environment}).  A command-line interface, the agave API (\url{http://agaveapi.co/}, \citep{dooley2012agave}), is linked to the Discovery environment. The Agave API is used for computationally intensive applications. The API is a RESTful application programming interface \citep{fielding2000} that allows direct interaction with all of iPlant resources.  The only way to access the API is to use {\it cURL} statements \citep{warranty1996}, an example of a {\it cURL} statement will be detailed in one of the sections.  The API provides access to authentication, data manipulation and storage, and job submittal via HTTPS- and command-line functions. The benefit of using the API is having programmatic access that allows advantages to power users (e.g. submitting jobs via batch files).  The {\it rPlant} package provides a direct link between high performance resources located at the Texas Advanced Computing Center (\url{http://www.tacc.utexas.edu}) that the API can access and the R environment, by essentially creating wrappers around the {\it cURL} statements, using the {\it RCurl} \citep{lang2007r} package. 

% also cite agave itself, http://agaveapi.co/citing-agave/  


\section{Getting Started}

\noindent This vignette assumes you have the current version of R \citep{R}.  First, install and load the package.  A stable release is available through CRAN (\url{http://cran.r-project.org/web/packages/rPlant/}) or a working repository can also be used through R-Forge (\url{https://r-forge.r-project.org/projects/rplant/}). \\

\noindent You can register as an iPlant user on their website (\url{http://user.iplantcollaborative.org/}) generating a unique username and password combination. 

 \subsection{Gaining Access to the API}

\noindent {\ttfamily Validate(username, password, api="agave")} \\

\noindent The username/password combination will be used in the {\ttfamily Validate} function.  The {\ttfamily Validate} function is required for every rPlant session and needs to be the first thing executed or the session will fail. In addition, it has a four-hour expiration. {\it rPlant} functions will auto renew a session, thereby extending the expiration. If a session sits idle and expires, the user will not have to re-validate a session as the functions will do this automatically.  The only time a user will need to use the {\ttfamily Validate}  function is at the start of a new session.  

% Modify the following if you are going to compile the document in Sweave:
% Then enter the code as commented out parts below...

{\fs
<<input, echo=TRUE, eval=FALSE>>=
require(rPlant)
username <- "enter your username"
password <- "enter your secret password"
Validate(username, password, api="agave")
@

<<inputres, echo=FALSE, eval=TRUE, results=hide>>=
require(rPlant)
require(ape)
# username <- "bleplant"
# password <- "Chl0r0phyll"
username <- "henryl"
password <- "Blizzard1"
tryCatch(Validate(username, password, api="foundation"), error=function(err){return(paste(err))})
tryCatch(Validate(username, password, api="foundation"), error=function(err){return(paste(err))})
@
}

\noindent The function checks if the username and password are valid iPlant credentials.  If they aren't the above error is displayed.  If the function is successful then nothing is printed. On {\it rPlant's} backend, a new R environment (rPlant.env) was created that stores all of the validation objects, including the user key and secret, user name and password, and token expiration. These items can be examined by using the  {\ttfamily ls(rPlant.env)} and using the \$ operator to display individual objects. \\

\noindent Every {\it rPlant} function has the option {\ttfamily print.curl=TRUE} or {\ttfamily FALSE}.  This refers to {\ttfamily cURL} a computer software project providing a way to transfer data using various protocols, for detail on {\ttfamily cURL} see \url{http://en.wikipedia.org/wiki/CURL}.  These statements (w/o the outside quotes) can be copied and pasted into a terminal in linux or unix.  And if {\ttfamily cURL} is installed on the computer then the statements can be executed.  You will see that these statements do the exact same thing as the {\it rPlant} functions.  This is one of the big advantages of {\it rPlant}, it can be used on any computer (including windows) and there is no need for the user to install {\ttfamily cURL} on that computer, because {\it rPlant} uses the package {\ttfamily RCurl}. \\

\noindent {\it Note}: This package abides by the unix rule, ``silence is golden".  If a function is successful then no output will be displayed.  If an error is attained then the error will be printed.

\section{Uploading Files}
\subsection{UploadFile function}

\noindent {\ttfamily UploadFile(local.file.name, local.file.path="", filetype=NULL, print.curl=FALSE, suppress.Warnings=FALSE)} \\

\noindent The first step is to upload files onto iPlants. {\ttfamily UploadFile} takes a file from your computer and uploads it onto iPlants servers, it does NOT take a file from the R workspace. Objects in the workspace will need to be saved to the computer in a supported file format before uploading them. 

{\fs
<<uploadfiledna, echo=TRUE, eval=FALSE>>=
data(DNA.fasta)
write.fasta(sequences = DNA.fasta, names = names(DNA.fasta), file.out = "DNA.fasta")
UploadFile(local.file.name="DNA.fasta")
@

<<uploadfilednares, echo=FALSE, eval=TRUE, results=verbatim>>=
data(DNA.fasta)
write.fasta(sequences = DNA.fasta, names = names(DNA.fasta), file.out = "DNA.fasta")
tryCatch(UploadFile(local.file.name="DNA.fasta"), error=function(err){return(paste(err))})
@
}

\noindent In the event that the file already exists on the iPlant server, an error will report and the file will not upload. For details on how to check contents of iPlant directories, see Section 4, and for manipulating files (like deleting, moving, sharing) see Section 5 below.  

\noindent The file format for the uploaded file can be defined using {\ttfamily filetype}, some programs will only accept certain types of files.  This can also be left as NULL, if the iPlant application doesn't require it. For the fasta file the file type is FASTA-0.  See the following section for the various file types supported by the API. 

\subsection{Supported File Types}

\noindent {\ttfamily SupportFile(print.curl=FALSE)} \\

\noindent There are 33 other file types supported by iPlant, use the {\ttfamily SupportFile} function to see all of the available file types (i.e. PHYLIP file type is ``PHYLIP-0" and ClustalW is ``ClustalW-1.8").

{\fs
<<supportfile, echo=TRUE>>=
SupportFile()
@
}

\section{Manipulating directories on iPlant servers}

\noindent Now that the file DNA.fasta has been uploaded onto the iPlant servers we can look at the file (or at least see which directory the file is in) by using the {\ttfamily ListDir} function.  There are a few other directory manipulation functions, they are: {\ttfamily MakeDir}, {\ttfamily ShareDir}, {\ttfamily PermissionsDir}, {\ttfamily RenameDir}, {\ttfamily MoveDir} and {\ttfamily DeleteDir}.

\subsection{Listing directories}

\noindent {\ttfamily ListDir(dir.name, dir.path="", print.curl=FALSE, shared.username=NULL, \\
\noindent suppress.Warnings=FALSE)} \\

\noindent Looking in the home directory you can see the ``DNA.fasta'' file.

{\fs
<<ListDir, echo=TRUE>>=
ListDir(dir.name="", suppress.Warnings=TRUE)
@
}

\noindent {\it Note}: Some functions contain an option, {\ttfamily suppress.Warnings}.  If you are absolutely sure that the commands you are entering are correct then to speed up the process have {\ttfamily suppress.Warnings=TRUE}.  But be careful, if used inappropriately then files could get overwritten or the files might not even exist and you will get no warning about it. %Probably a good idea to add a sentence or two about what types of warnings are being suppressed.


\noindent {\ttfamily MakeDir(dir.name, dir.path="", print.curl=FALSE, suppress.Warnings=FALSE)} \\

\noindent The following function will make a directory hello in the home directory.

<<DeleteDirHidden, echo=FALSE, results=HIDE>>=
tryCatch(DeleteDir(dir.name="hello"), error=function(err){return(paste(err))})
@

{\fs
<<MakeDirhello, echo=TRUE>>=
MakeDir(dir.name="hello")
@
}

\noindent Again making the directoy ``all" in the ``hello" directory.

{\fs
<<MakeDirall, echo=TRUE>>=
MakeDir(dir.name="all", dir.path="hello")
@
}


\noindent Here we make another directory ``robots'' in the ``all'' directory.  Note how the {\ttfamily dir.path} needs to be constructed, and how the {\ttfamily dir.name} and {\ttfamily dir.path} are related.  All of the functions have this same format.

{\fs
<<MakeDirrobots, echo=TRUE>>=
MakeDir(dir.name="robots", dir.path="hello/all")
@
}

\noindent We can look inside the ``hello/all/robots'' directory and see that there is nothing in there.

{\fs
<<ListDirrobots, echo=TRUE>>=
ListDir(dir.name="robots", dir.path="hello/all")
@
}

\subsection{Sharing Directories}

\noindent {\ttfamily ShareDir(dir.name, dir.path="", shared.username, read=TRUE, execute=TRUE, \\
\noindent print.curl=FALSE, suppress.Warnings=FALSE)} \\

\noindent A really nice feature of {\it iPlant} is the file sharing feature.  As was said in the introduction one of iPlant's goals was to work with very large data sets.  And when data sets are too large to send via email then a sharing feature is absolutely necessary.  There are in fact two share functions, one for sharing a single file ({\ttfamily ShareFile}) and the other ({\ttfamily ShareDir}) for sharing an entire directory. When sharing a directory, all files contained within will be shared. \\  %is there an upper limit on file size?

\noindent In this sample we share the ``all'' directory.  

{\fs
<<ShareDir, echo=TRUE>>=
ShareDir(dir.name="all", dir.path="hello", shared.username="kamichels")
@
}

\noindent In the above example I share something with user ``henryl''. I can also view files and directories that have been shared with me.

{\fs
<<ListDirShareDir, echo=TRUE>>=
ListDir(dir.name="data", dir.path="", shared.username="phyllisl")
@
}

\noindent Several other functions use the {\ttfamily shared.username} option. For example, the {\ttfamily SubmitJob} function and the wrapper functions.  A job can be run on files that are shared with you (see Section 7).

\subsection{Checking Permissions on Directories}

\noindent {\ttfamily PermissionsDir(dir.name, dir.path="", print.curl=FALSE, suppress.Warnings=FALSE)} \\

\noindent When using the sharing feature of rPlant, a user needs a function which will check the permissions of a directory.  This includes the users with whom the directory is shared and what the permissions are included.  Below confirms the sharing performed earlier with ``kamichels''.

{\fs
<<PermsDir, echo=TRUE>>=
PermissionsDir(dir.name="all", dir.path="hello")
@
}

\subsection{Renaming Directories}

\noindent {\ttfamily RenameDir(dir.name, new.dir.name, dir.path="", print.curl=FALSE, \\
\noindent suppress.Warnings=FALSE)} \\

\noindent This function renames a directory.

{\fs
<<RenameDir, echo=TRUE>>=
RenameDir("robots", "tools", "hello/all")
@
}

\noindent And you can see that it has been changed. \\

{\fs
<<ListDirRenameDir, echo=TRUE>>=
ListDir("all", "hello")
@
}

\noindent {\it Note}: When the directory is renamed, permissions on sharing will have to be redone.

\subsection{Moving Directories}

\noindent {\ttfamily MoveDir(dir.name, dir.path="", end.path="", print.curl=FALSE, \\
\noindent suppress.Warnings=FALSE)} \\

\noindent This function moves a directory. The following code will move the directory 'tools' from 'hello/all' to the home directory.  Verified below.

<<DeleteDirHidden2, echo=FALSE, results=HIDE>>=
tryCatch(DeleteDir(dir.name="tools"), error=function(err){return(paste(err))})
@

{\fs
<<MoveDir, echo=TRUE>>=
MoveDir("tools", "hello/all", end.path="")
ListDir("")
@
}

\noindent {\it Note}: When the directory is moved, permissions on sharing will have to be redone.

\subsection{Deleting Directories}

\noindent {\ttfamily DeleteDir(dir.name, dir.path="", print.curl=FALSE, suppress.Warnings=FALSE)} \\

\noindent This function deletes a directory and all of the subdirectories.

{\fs
<<DeleteDir, echo=TRUE>>=
DeleteDir("tools")
ListDir("")
@
}

\section{Manipulating files on iPlant servers}

\noindent The file manipulation tools available in this package are very similar to directory manipulation tools.  The file manipulation functions are: {\ttfamily ShareFile}, {\ttfamily PermissionsFile}, {\ttfamily RenameFile}, {\ttfamily MoveFile} and {\ttfamily DeleteFile}.

\subsection{Sharing Files}

\noindent {\ttfamily ShareFile(file.name, file.path="", shared.username, read=TRUE, execute=TRUE, \\
\noindent print.curl=FALSE, suppress.Warnings=FALSE)} \\

\noindent As described in the {\ttfamily ShareDir} function, a really nice feature of iPlant is the file sharing feature. This function shares one file at a time. \\

\noindent The following code will share the file DNA.fasta with phyllisl.

{\fs
<<ShareFile, echo=TRUE>>=
ShareFile(file.name="DNA.fasta", shared.username="phyllisl")
@
}

\subsection{Checking Permissions on a File}

\noindent {\ttfamily PermissionsFile(file.name, file.path="", print.curl=FALSE, suppress.Warnings=FALSE)} \\

\noindent When using the sharing feature of rPlant, a user needs a function which will check the permissions of a file.  This includes the users with whom the file is shared and what the permissions are included.  Below confirms the sharing performed earlier with ``phyllisl''.

{\fs
<<PermsFile, echo=TRUE>>=
PermissionsFile(file.name="DNA.fasta")
@
}

\subsection{Moving Files}

\noindent {\ttfamily MoveFile(file.name, file.path="", end.path="", print.curl=FALSE, \\
\noindent suppress.Warnings=FALSE)} \\

\noindent This function moves the file from one directory to another.

{\fs
<<MoveFile, echo=TRUE>>=
MoveFile("DNA.fasta", end.path="hello/all")
ListDir("all", "hello")
@
}

\noindent The move took the file DNA.fasta from the home directory into the ``hello/all'' directory. \\

\noindent {\it Note}: When the file is moved, permissions on sharing will have to be redone.

\subsection{Renaming Files}

\noindent {\ttfamily RenameFile(file.name, new.file.name, file.path="", print.curl=FALSE, \\
\noindent suppress.Warnings=FALSE)} \\

\noindent This function renames a single file.

{\fs
<<RenameFile, echo=TRUE>>=
RenameFile("DNA.fasta", "lp.fasta", "hello/all")
ListDir("all", "hello")
@
}

\noindent {\it Note}: When the file is renamed, permissions on sharing will have to be redone.

\subsection{Deleting Files}

\noindent {\ttfamily DeleteFile(file.name, file.path="", print.curl=FALSE, suppress.Warnings=FALSE)} \\

\noindent This function deletes a single file in the specified directory.

{\fs
<<DeleteFile, echo=TRUE>>=
DeleteFile("lp.fasta", "hello/all")
ListDir("all", "hello")
@
}

\noindent The file lp.fasta is no longer in the ``hello/all" directory. \\

\section{Applications}

\noindent The real power in the {\it rPlant} package is to have the programmatic access to the phylogenetic tools/applications that are available in the API.  {\it rPlant} can be used to interact with any of the API applications. 

\subsection{Listing Applications}

\noindent {\ttfamily ListApps(description=FALSE, print.curl=FALSE)} \\

\noindent This function returns a sorted list of the newest versions of the public applications that are available via the Foundation API.  These applications are ones that can be used with the {\ttfamily SubmitJob} function (see Section 7).

{\fs
<<ListApps, echo=TRUE>>=
ListApps(description=FALSE)
@
}

\noindent {\it Note}:  As said above, these applications are PUBLIC applications. Applications in the API are split into two categories, public and private.  Private applications are ones that are developed and tested and changed.  Only the user who created the private application can use it.  The other category is public applications.  After a private application has gone through extensive testing, then the application can be published and it becomes a public application which is available to all iPlant users.  In the API a public application is labeled by adding the suffix 'u1' to it.  The '1' is referred to as the version number, so if a public application is fixed and republished the suffix becomes 'u2'.

\subsection{Application Information}

\noindent {\ttfamily GetAppInfo(application, return.json=FALSE, print.curl=FALSE)}\\

\noindent The {\ttfamily GetAppInfo} function returns the application with a short description and the input/output filetypes.

{\fs
<<GetAppInfo, echo=TRUE>>=
GetAppInfo("velveth-1.2.07u1")
@
}

\noindent The {\ttfamily GetAppInfo} function returns a list of information about the application that is needed for the {\ttfamily SubmitJob} function. The first element gives a short description of the application. The second element in the list gives information on the application including its use permissions (public vs. private) and whether it is the newest version. The third element in the list  is a matrix with four columns of information: kind, id, file type or value, and any details. In the example above, first column ('kind') states there are six inputs for this app, the 'id' column names those inputs as 'reads5', 'reads3', etc. There are also eight parameters for the app, such as 'format2', 'kmer', etc. The third column ('Type/value') returns the type of file the application is expecting if it is input or it returns the type of input necessary for the application parameters, common ones are string, boolean, etc. The last column gives brief details on each input.

\section{Submitting Jobs in the rPlant package}

\subsection{Submitting Job}

\noindent {\ttfamily SubmitJob(application, file.path="", file.list=NULL, input.list, args.list=NULL, job.name, nprocs=1, private.APP=FALSE, suppress.Warnings=FALSE, shared.username=NULL, print.curl=FALSE, email=TRUE)}\\

\noindent An important benefit of using {\it rPlant} is the ability to create batch-scripted files that automate job submittal and retrieval. For example, a user could submit parallel alignment jobs of different gene regions or multiple jobs with the same data and different parameter values. The results could then be automatically downloaded upon completion. \\

\noindent The following function is the main way to submit a job to the iPlant server, and can be used for any iPlant application. You can also submit jobs via the wrapper functions (for example, {\ttfamily Muscle()}, see Section 9), which call upon the {\ttfamily SubmitJob} function internally.  

<<UploadFileR, echo=FALSE, results=HIDE>>=
tryCatch(DeleteJob(ALL=TRUE), error=function(err){return(paste(err))})
data(DNA.fasta)
write.fasta(sequences = DNA.fasta, names = names(DNA.fasta), file.out = "DNA.fasta")
tryCatch(UploadFile(local.file.name="DNA.fasta"), error=function(err){return(paste(err))})
@

{\fs
<<SubmitJobReg, echo=TRUE, eval=FALSE>>=
UploadFile(local.file.name="DNA.fasta", filetype="FASTA-0")
ListDir("")
myJobM <- SubmitJob(application="Muscle-3.8.32u4", file.list=list("DNA.fasta"),
                    input.list=list("stdin"), args.list=list(c("arguments", 
                    "-phyiout -center -cluster1 upgma")), job.name="Muscle")
@

<<SubmitJobRegres, echo=FALSE, eval=TRUE, results=verbatim>>=
ListDir("")
myJobM <- SubmitJob(application="Muscle-3.8.32u4", file.list=list("DNA.fasta"),
                    input.list=list("stdin"), args.list=list(c("arguments", 
                    "-phyiout -center -cluster1 upgma")), job.name="Muscle")
@
}

\noindent Several important argument definitions are listed below, but can also be found in the help files: 
\noindent {\ttfamily input.list}:  This argument defines what kind of input you are passing the application. You can get application information from the {\ttfamily GetAppInfo} function ({\ttfamily GetAppInfo("Muscle-3.8.32u4")\$Information}).  In this example, the 'kind' column states there is one input for this app, and the 'id' column names that input as 'stdin'.  Input types change from application to application. \\

\noindent {\ttfamily file.list}: Similar to {\ttfamily input.list}, the {\ttfamily file.list} argument defines which files are being passed to the application. The named file must be on the DE within the {\ttfamily file.path} and be formatted to the correct specification (for example, {\ttfamily GetAppInfo("Muscle-3.8.32u4")\$Information}). If it the file types don't match then the application will fail. \\

\noindent {\ttfamily args.list}:  The {\ttfamily args.list} is where application flagging options can be entered. These typically change default options. Using information from the {\ttfamily GetAppInfo} function, the 'kind' column states there is one parameter for this app, the 'id' column gives the name of that parameter is ``arguments", and the  ``fileType/value" column tells me it is a string. This is where the fourth column 'details' comes in handy; it tells me that the parameter input is ``program arguments and options", which means it can accept a string. For example: {\ttfamily args.list=list(c(arguments, "-phyiout -center -cluster1 upgma"))} \\

\noindent The {\ttfamily args.list} is a list that is as long as the number of parameters (so length 1 in the Muscle example), that means there as many vectors as there are parameters. All vectors are of the same length, two in this example. In the first position, is the name of the parameter, ``arguments", and in the second position is the value of that parameter, ``-phyiout -center -cluster1 upgma''. In this example, it is a string of command line flags. \\

\noindent The function {\ttfamily SubmitJob} will return a list of two objects (in this example, {\ttfamily myJobM}). The first object, is the job number and the second is the job name, both are important information for retrieving results. \\

\noindent If job submittal is successful, then the function automatically creates the folder ``analyses" within a user's cloud (if it did not previously exist). If the job finishes, then a folder is created within the analyses folder that is named the job name.

\subsection{Submitting a job with a shared file}

\noindent Jobs can also be submitted from files stored in other user's cloudspace that are shared.  In the below example, a file that had been previously shared (Section 3), a job can be submitted using that file.

{\fs
<<SubmitJobShare, echo=TRUE>>=
myJobS <- SubmitJob(application="Muscle-3.8.32u4", file.list=list("muscle3.fasta"), 
                    file.path="data", shared.username="phyllisl", 
                    args.list=list(c("arguments", "-fastaout")), 
                    input.list=list("stdin"), job.name="MuscleShare")
@
}

\section{Checking Job Status and Retrieving Job output}

\noindent Once the job is submitted, it is assigned a job identification number (job.id) that we can use to check the status and download any results files. The job.id is returned with the SubmitJob function, so if you create an object when you submit a job then you can use that object as an identifier as well.  Otherwise, you can copy/paste a job ID into any of the functions as a character string.  
The job number is used in a few rPlant functions including, {\ttfamily CheckJobStatus}, {\ttfamily KillJob}, {\ttfamily ListJobOutput}, {\ttfamily RetrieveJob} and {\ttfamily DeleteJob}. If you need to get job IDs from older jobs, you can use the function {\ttfamily GetJobHistory()}. 

\subsection{Checking job status}

\noindent {\ttfamily CheckJobStatus(job.id, print.curl=FALSE)} \\

\noindent This function checks the status of a job on the iPlant servers.

\begin{table}[ht]
\caption{Possible Outputs for {\ttfamily CheckJobStatus()}}
% title of Table
\centering
% used for centering table
\begin{tabular}{l}
% centered columns (1 columns)
\hline\hline
%inserts double horizontal lines
{\bfseries Stages} \\ [0.5ex]
% inserts table
%heading
\hline
% inserts single horizontal line
{\ttfamily PENDING}            \\
{\ttfamily STAGING\_INPUTS}     \\
{\ttfamily CLEANING\_UP}        \\
{\ttfamily ARCHIVING}          \\
{\ttfamily STAGING\_JOB}        \\
{\ttfamily FINISHED}           \\
{\ttfamily KILLED}             \\
{\ttfamily FAILED}             \\
{\ttfamily STOPPED}            \\
{\ttfamily RUNNING}            \\
{\ttfamily PAUSED}             \\
{\ttfamily QUEUED}             \\
{\ttfamily SUBMITTING}         \\
{\ttfamily STAGED}             \\
{\ttfamily PROCESSING\_INPUTS}  \\
{\ttfamily ARCHIVING\_FINISHED} \\
{\ttfamily ARCHIVING\_FAILED}  \\[1ex]
% [1ex] adds vertical space
\hline
%inserts single line
\end{tabular}
\label{table:nonlin}
% is used to refer this table in the text
\end{table}

{\fs
<<CheckJobStatus, echo=TRUE>>=
CheckJobStatus(myJobS[[1]])
@
}

\subsection{Killing Job}

\noindent If you don't want a job to run any more, then simply stop it by using the {\ttfamily KillJob} function.  Generally this is used if a job is {\ttfamily QUEUED}, it must be stopped first to delete it.

{\fs
<<KillJob, echo=TRUE>>=
KillJob(myJobM[[1]])
@
}

\subsection{Listing job status}

\noindent {\ttfamily ListJobOutput(job.id, print.curl=FALSE, print.total=TRUE)} \\

\noindent This function lists the output files from a finished job. For example, these files are the output from our MUSCLE example above.  

%% I really need to explain the Wait function, it is part of rPlant, don't need while loop.
<<Wait, result=HIDE, echo=FALSE>>=
minWait = 5 # seconds
maxWait = 1800 # 30 min in seconds

Wait(myJobS[[1]], minWait, maxWait)
@

{\fs
<<ListJobOutput, echo=TRUE, results=verbatim>>=
ListJobOutput(myJobS[[1]])
@
}

\subsection{Looking at Job History}

\noindent {\ttfamily GetJobHistory(return.json=FALSE, print.curl=FALSE)} \\

\noindent This function displays the entire job history for the user. This is an easy way to grab old job IDs in order to retrieve files or check the status of a set of jobs.  You can see that one finished and one was stopped.  

{\fs
<<GetJobHistory, echo=TRUE>>=
GetJobHistory()
@
}

\subsection{Retrieve job files}

\noindent {\ttfamily RetrieveJob(job.id, file.vec=NULL, print.curl=FALSE, verbose=FALSE)} \\

\noindent One very handy thing about the {\it rPlant} package is the ability to download the files directly from the iPlant servers to your computer.  The following downloads all of the output files.

{\fs
<<RetrieveJob, echo=TRUE>>=
RetrieveJob(myJobS[[1]])
@
}

\noindent The files have been downloaded into a new directory within your working directory. If you want to download select files at a time, then these will need to be defined within the file.vec argument (for example: {\ttfamily RetrieveJob(myJobM, file.vec=c("fasta.aln"))}) 

\subsection{Delete job}

\noindent {\ttfamily DeleteJob(job.id, print.curl=FALSE, ALL=FALSE)} \\

\noindent After the job has been submitted, the results downloaded, and you have no need for the job anymore, you can use the {\ttfamily DeleteJob} function to delete the job. The nice thing about this function is that not only will it delete the job number from the job history but it will also delete the job folder and all contents in the analyses folder in the user's cloudspace.

{\fs
<<DeleteJob, echo=TRUE>>=
DeleteJob(myJobS[[1]])
@
}

\noindent You also have the option to erase ALL job history from a user's past.  This may be useful after a round of testing.  

{\fs
<<DeleteJobAll, echo=TRUE, eval=FALSE>>=
DeleteJob(ALL=TRUE)
@

<<DeleteJobAllres, echo=FALSE, eval=TRUE, results=verbatim>>=
tryCatch(DeleteJob(ALL=TRUE), error=function(err){return(paste(err))})
@
}

\section{Submitting Jobs With Wrappers}

\noindent We have ten dedicated wrapper functions for iPlant applications that will ease submitting jobs. These wrappers will use many application defaults and/or change flags into wrapper function arguments.  Of course, if a user needs more flexibility in flagging options, then they can still submit jobs using the {\ttfamily SubmitJob} function. \\

\noindent Of the iPlant applications that have dedicated wrappers, there is a clear bias towards phylogenetic applications because the authors are evolutionary biologists and regularly use these programs. Writing wrapper functions is not programmatically difficult, but it does require familiarity with the individual programs and their associated data sets. We would like to encourage any users who are using programs without wrappers to submit patches adding wrapper functions or request to be a developer. You can make these feature requests at the R-Forge site: \url{https://r-forge.r-project.org/tracker/?group_id=1328}. \\

\noindent Among the wrappers there are three which do alignments: {\ttfamily Muscle}, {\ttfamily Mafft} and {\ttfamily ClustalW}.  The alignments will do both protein and nucleotide.  Also make sure that the taxon names in the sequence files do not contain tabulators, carriage returns, spaces, ":", ",", ")", "(", ";", "]", "["''.

{\fs
<<uploadfileprotein, echo=TRUE, eval=FALSE>>=
data(PROTEIN.fasta)
write.fasta(sequences=PROTEIN.fasta, names=names(PROTEIN.fasta), file.out="PROTEIN.fasta")
UploadFile(local.file.name="PROTEIN.fasta", filetype="FASTA-0")
@

<<uploadfileproteinres, echo=FALSE, eval=TRUE, results=verbatim>>=
data(PROTEIN.fasta)
write.fasta(sequences=PROTEIN.fasta, names=names(PROTEIN.fasta), file.out="PROTEIN.fasta")
tryCatch(UploadFile(local.file.name="PROTEIN.fasta", filetype="FASTA-0"), error=function(err){return(paste(err))})
@
}

\subsection{Muscle}

\noindent {\ttfamily Muscle(file.name, file.path="", job.name=NULL, args=NULL, version="Muscle-3.8.32u4", print.curl=FALSE, aln.filetype="PHYLIP\_INT", shared.username=NULL, \\
\noindent suppress.Warnings=FALSE)} \\

\noindent MUSCLE is a program for creating multiple alignments of amino acid or nucleotide sequences. A range of options is provided that give you the choice of optimizing accuracy, speed, or some compromise between the two.  The manual is also available here: \url{http://www.drive5.com/muscle/muscle\_userguide3.8.html}

<<>>=
myJobMuDP <- Muscle("DNA.fasta", aln.filetype="PHYLIP_INT", job.name="muscleDNAphyINT")
myJobMuDF <- Muscle("DNA.fasta", aln.filetype="FASTA", job.name="muscleDNAfasta")
myJobMuDPP <- Muscle("DNA.fasta", aln.filetype="PHYLIP_PARS", job.name="muscleDNAphyPARS")
myJobMuDPS <- Muscle("DNA.fasta", aln.filetype="PHYLIP_SEQ", job.name="muscleDNAphySEQ")
myJobMuDC <- Muscle("DNA.fasta", aln.filetype="CLUSTALW", job.name="muscleDNAclustalw")
myJobMuDM <- Muscle("DNA.fasta", aln.filetype="MSF", job.name="muscleDNAmsf")
myJobMuPP <- Muscle("PROTEIN.fasta", aln.filetype="PHYLIP_INT", job.name="musclePROTEINphyINT")
myJobMuPF <- Muscle("PROTEIN.fasta", aln.filetype="FASTA", job.name="musclePROTEINfasta")
myJobMuPPP <- Muscle("PROTEIN.fasta", aln.filetype="PHYLIP_PARS", job.name="musclePROTEINphyPARS")
myJobMuPPS <- Muscle("PROTEIN.fasta", aln.filetype="PHYLIP_SEQ", job.name="musclePROTEINphySEQ")
myJobMuPC <- Muscle("PROTEIN.fasta", aln.filetype="CLUSTALW", job.name="musclePROTEINclustalw")
myJobMuPM <- Muscle("PROTEIN.fasta", aln.filetype="MSF", job.name="muscleDNAmsf")
@

\noindent MUSCLE outputs six alignments: fasta.aln (\url{http://en.wikipedia.org/wiki/FASTA\_format}), phylip\_sequential.aln, phylip\_interleaved.aln, phylip\_pars.aln (\url{http://www.bioperl.org/wiki/PHYLIP\_multiple\_alignment\_format}), clustalw.aln (\url{http://meme.nbcr.net/meme/doc/clustalw-format.html}) and msf.aln (\url{http://en.wikipedia.org/wiki/MSF}).

\subsection{Mafft}

\noindent {\ttfamily Mafft(file.name, file.path="", type="DNA",  print.curl=FALSE, version="mafftDispatcher-1.0.13100u1", args=NULL, job.name=NULL, aln.filetype="FASTA", shared.username=NULL, suppress.Warnings=FALSE)} \\

\noindent MAFFT is a multiple sequence alignment program for unix-like operating systems.  It offers a range of multiple alignment methods, L-INS-i (accurate; for alignment of about 200 sequences), FFT-NS-2 (fast; for alignment of about 10,000 sequences), etc.  See \url{http://mafft.cbrc.jp/alignment/software/}.  The manual is also available here: \url{http://mafft.cbrc.jp/alignment/software/manual/manual.html}.

<<>>=
myJobMaDF <- Mafft("DNA.fasta", job.name="mafftDNAfasta")
myJobMaDC <- Mafft("DNA.fasta", aln.filetype="CLUSTALW", job.name="mafftDNAclustalw")
myJobMaPF <- Mafft("PROTEIN.fasta", type="PROTEIN", job.name="mafftPROTEINfasta")
myJobMaPC <- Mafft("PROTEIN.fasta", type="PROTEIN", aln.filetype="CLUSTALW",
                    job.name="mafftPROTEINclustalw")
@

\noindent MAFFT outputs two alignments (both named: mafft.fa): {\ttfamily FASTA} (\url{http://en.wikipedia.org/wiki/FASTA\_format}) and {\ttfamily CLUSTALW} (\url{http://meme.nbcr.net/meme/doc/clustalw-format.html}).

\subsection{ClustalW}

\noindent {\ttfamily ClustalW(file.name, file.path="", type="DNA", job.name=NULL, version="ClustalW2-2.1u1", print.curl=FALSE, args=NULL, aln.filetype="PHYLIP", shared.username=NULL, suppress.Warnings=FALSE)} \\

\noindent An approach for performing multiple alignments of large numbers of amino acid or nucleotide sequences is described. The method is based on first deriving a phylogenetic tree from a matrix of all pairwise sequence similarity scores, obtained using a fast pairwise alignment algorithm.  See details on \url{http://www.clustal.org/clustal2/}.

<<>>=
myJobCWDP <- ClustalW("DNA.fasta", job.name="clustalwDNAphylip")
myJobCWDC <- ClustalW("DNA.fasta", aln.filetype="CLUSTALW", job.name="clustalwDNAclustalw")
myJobCWDN <- ClustalW("DNA.fasta", aln.filetype="NEXUS", job.name="clustalwDNAnexus")
myJobCWDGCG <- ClustalW("DNA.fasta", aln.filetype="GCG", job.name="clustalwDNAgcg")
myJobCWDGDE <- ClustalW("DNA.fasta", aln.filetype="GDE", job.name="clustalwDNAgde")
myJobCWDPIR <- ClustalW("DNA.fasta", aln.filetype="PIR", job.name="clustalwDNApir")
myJobCWPP <- ClustalW("PROTEIN.fasta", type="PROTEIN", job.name="clustalwPROTEINphylip")
myJobCWPC <- ClustalW("PROTEIN.fasta", type="PROTEIN", aln.filetype="CLUSTALW",
                       job.name="clustalwPROTEINclustalw")
myJobCWPN <- ClustalW("PROTEIN.fasta", type="PROTEIN", aln.filetype="NEXUS",
                       job.name="clustalwPROTEINnexus")
myJobCWPGCG <- ClustalW("PROTEIN.fasta", type="PROTEIN", aln.filetype="GCG",
                       job.name="clustalwPROTEINgcg")
myJobCWPGDE <- ClustalW("PROTEIN.fasta", type="PROTEIN", aln.filetype="GDE",
                       job.name="clustalwPROTEINgde")
myJobCWPPIR <- ClustalW("PROTEIN.fasta", type="PROTEIN", aln.filetype="PIR",
                       job.name="clustalwPROTEINpir")
@

\noindent {\ttfamily ClustalW} outputs six alignments (all named: clustalw.fa): CLUSTALW \url{http://meme.nbcr.net/meme/doc/clustalw-format.html}, PHYLIP\_INT \url{http://www.bioperl.org/wiki/PHYLIP\_multiple\_alignment\_format}, NEXUS \url{http://en.wikipedia.org/wiki/Nexus\_file}, GCG \url{http://www.genomatix.de/online\_help/help/sequence\_formats.html#GCG}, GDE \url{http://www.cse.unsw.edu.au/~binftools/birch/GDE/overview/GDE.file\_formats.html},  and PIR \url{http://www.bioinformatics.nl/tools/crab\_pir.html}.

<<WaitAll1, results=HIDE, echo=FALSE>>=
Wait(myJobMuDP[[1]], minWait, maxWait)
Wait(myJobMuDF[[1]], minWait, maxWait)
Wait(myJobMuDPP[[1]], minWait, maxWait)
Wait(myJobMuDPS[[1]], minWait, maxWait)
Wait(myJobMuDC[[1]], minWait, maxWait)
Wait(myJobMuDM[[1]], minWait, maxWait)
Wait(myJobMuPP[[1]], minWait, maxWait)
Wait(myJobMuPF[[1]], minWait, maxWait)
Wait(myJobMuPPP[[1]], minWait, maxWait)
Wait(myJobMuPPS[[1]], minWait, maxWait)
Wait(myJobMuPC[[1]], minWait, maxWait)
Wait(myJobMuPM[[1]], minWait, maxWait)
Wait(myJobMaDF[[1]], minWait, maxWait)
Wait(myJobMaDC[[1]], minWait, maxWait)
Wait(myJobMaPF[[1]], minWait, maxWait)
Wait(myJobMaPC[[1]], minWait, maxWait)
Wait(myJobCWDP[[1]], minWait, maxWait)
Wait(myJobCWDC[[1]], minWait, maxWait)
Wait(myJobCWDN[[1]], minWait, maxWait)
Wait(myJobCWDGCG[[1]], minWait, maxWait)
Wait(myJobCWDGDE[[1]], minWait, maxWait)
Wait(myJobCWDPIR[[1]], minWait, maxWait)
Wait(myJobCWPP[[1]], minWait, maxWait)
Wait(myJobCWPC[[1]], minWait, maxWait)
Wait(myJobCWPN[[1]], minWait, maxWait)
Wait(myJobCWPGCG[[1]], minWait, maxWait)
Wait(myJobCWPGDE[[1]], minWait, maxWait)
Wait(myJobCWPPIR[[1]], minWait, maxWait)
@

\subsection{FastTree}

\noindent {\ttfamily Fasttree <- function(file.name, file.path="", job.name=NULL, args=NULL, type="DNA", model=NULL, gamma=FALSE, stat=FALSE, print.curl=FALSE, version="fasttreeDispatcher-1.0.0u1", shared.username=NULL, suppress.Warnings=FALSE)} \\

\noindent {\ttfamily FastTree} infers approximately-maximum-likelihood phylogenetic trees from alignments of nucleotide or protein sequences. See \url{http://meta.microbesonline.org/fasttree/}

{\fs
<<FasttreeDNA, echo=TRUE>>=
myJobFaDMuP <- Fasttree("phylip_interleaved.aln", file.path=paste("analyses/",myJobMuDP[[2]], 
                        sep=""), job.name="fasttreeMUSCLEdnaPHY")
myJobFaDCWP <- Fasttree("clustalw2.fa", file.path=paste("analyses/",myJobCWDP[[2]], sep=""), 
                        job.name="fasttreeCLUSTALWdnaPHY")
myJobFaDMuF <- Fasttree("fasta.aln", file.path=paste("analyses/",myJobMuDF[[2]], sep=""), 
                        job.name="fasttreeMUSCLEdnaFASTA")
myJobFaDCWF <- Fasttree("mafft.fa", file.path=paste("analyses/",myJobMaDF[[2]], sep=""), 
                        job.name="fasttreeMAFFTdnaFASTA")
@
}

{\fs
<<FasttreeProtein, echo=TRUE>>=
myJobFaPMuP <- Fasttree("phylip_interleaved.aln", file.path=paste("analyses/",myJobMuPP[[2]], 
                       sep=""), type="PROTEIN", 
                       job.name="fasttreeMUSCLEproteinPHY")
myJobFaPCWP <- Fasttree("clustalw2.fa", type="PROTEIN", file.path=paste("analyses/",myJobCWPP[[2]], 
                        sep=""), job.name="fasttreeCLUSTALWproteinPHY")
myJobFaPMuF <- Fasttree("fasta.aln", file.path=paste("analyses/",myJobMuPF[[2]], sep=""), 
                        type="PROTEIN", job.name="fasttreeMUSCLEproteinFASTA")
myJobFaPCWF <- Fasttree("mafft.fa", file.path=paste("analyses/",myJobMaPF[[2]], sep=""), 
                        type="PROTEIN", job.name="fasttreeMAFFTproteinFASTA")
@
}

\noindent {\ttfamily Fasttree} outputs trees in Newick format \url{http://en.wikipedia.org/wiki/Newick_format}. The placement of the root is not biologically meaningful. The local support values are given as names for the internal nodes, and range from 0 to 1, not from 0 to 100 or 0 to 1,000. If all sequences are unique, then the tree will be fully resolved (the root will have three children and other internal nodes will have two children). If there are multiple sequences that are identical to each other, then there will be a multifurcation. Also, there are no support values for the parent nodes of redundant sequences. 

\subsection{RAxML (Randomized Accelerated Maximum Likelihood)}

\noindent {\ttfamily RAxML(file.name, file.path="", job.name=NULL, type="DNA", model=NULL, bootstrap=NULL, algorithm="d", multipleModelFileName=NULL, args=NULL, numcat=25, nprocs=12, \\
\noindent version="raxml-lonestar-7.2.8u1", print.curl=FALSE, shared.username=NULL, \\
\noindent substitution\_matrix=NULL, empirical.frequencies=FALSE, suppress.Warnings=FALSE) } \\

\noindent {\ttfamily RAxML} is a program for sequential and parallel Maximum Likelihood based inference of large phylogenetic tress.  It has originall been derived from from fastDNAml which in turn was derived from Joe Felsentein's dnaml which is part of the PHYLIP package. See \url{http://sco.h-its.org/exelixis/oldPage/RAxML-Manual.7.0.4.pdf} for details.

{\fs
<<RAxMLDNA, echo=TRUE>>=
myJobRDMuP <- RAxML("phylip_interleaved.aln", file.path=paste("analyses/",myJobMuDP[[2]], 
                    sep=""), job.name="raxmlMUSCLEdnaPHY")
myJobRDCWP <- RAxML("clustalw2.fa", file.path=paste("analyses/",myJobCWDP[[2]], sep=""), 
                    job.name="raxmlCLUSTALWdnaPHY")
@
}

{\fs
<<RAxMLProtein, echo=TRUE>>=
myJobRPMuP <- RAxML("phylip_interleaved.aln", file.path=paste("analyses/",myJobMuPP[[2]], 
                    sep=""), type="PROTEIN", job.name="raxmlMUSCLEproteinPHY")
myJobRPCWP <- RAxML("clustalw2.fa", file.path=paste("analyses/",myJobCWPP[[2]], sep=""), 
                    type="PROTEIN", job.name="raxmlCLUSTALWproteinPHY")
@
}

\noindent For this application there are numerous output files.  See pg 16-17 of the manual for complete details.  RAxML outputs trees in Newick format \url{http://en.wikipedia.org/wiki/Newick\_format}.

\subsection{PHYLIP-Parsimony 3.69}

\noindent {\ttfamily PHYLIP\_Pars(file.name, file.path="", job.name=NULL, type="DNA", print.curl=FALSE, shared.username=NULL, suppress.Warnings=FALSE) } \\

\noindent {\ttfamily PHYLIP} is a free package of programs for inferring phylogenies. It is distributed as source code, documentation files, and a number of different types of executables. The web page: \url{http://evolution.genetics.washington.edu/phylip/doc/main.html}, by Joe Felsenstein of the Department of Genome Sciences and the Department of Biology at the University of Washington, contain information on PHYLIP.  PHYLIP (the PHYLogeny Inference Package) is a package of programs for inferring phylogenies (evolutionary trees).  Methods that are available in the package include parsimony, distance matrix, and likelihood methods, including bootstrapping and consensus trees.

{\fs
<<phylipDNA, echo=TRUE>>=
myJobPDMuPP <- PHYLIP_Pars("phylip_pars.aln", file.path=paste("analyses/",myJobMuDPP[[2]], 
                           sep=""), job.name="phylipMUSCLEdnaPHYpars")
@
}

{\fs
<<phylipProtein, echo=TRUE>>=
myJobPPMuPP <- PHYLIP_Pars("phylip_pars.aln", file.path=paste("analyses/",myJobMuPPP[[2]], 
                           sep=""), type="PROTEIN", job.name="phylipMUSCLEproteinPHYpars")
@
}

\noindent {\ttfamily PHYLIP Parsimon}y outputs trees in Newick format \url{http://en.wikipedia.org/wiki/Newick_format}.

<<WaitAll2, results=HIDE, echo=FALSE>>=
Wait(myJobFaDMuP[[1]], minWait, maxWait)
Wait(myJobFaDCWP[[1]], minWait, maxWait)
Wait(myJobFaDMuF[[1]], minWait, maxWait)
Wait(myJobFaDCWF[[1]], minWait, maxWait)
Wait(myJobFaPMuP[[1]], minWait, maxWait)
Wait(myJobFaPCWP[[1]], minWait, maxWait)
Wait(myJobFaPMuF[[1]], minWait, maxWait)
Wait(myJobFaPCWF[[1]], minWait, maxWait)
Wait(myJobRDMuP[[1]], minWait, maxWait)
Wait(myJobRDCWP[[1]], minWait, maxWait)
Wait(myJobRDMuP[[1]], minWait, maxWait)
Wait(myJobRDCWP[[1]], minWait, maxWait)
Wait(myJobRPMuP[[1]], minWait, maxWait)
Wait(myJobRPCWP[[1]], minWait, maxWait)
Wait(myJobPDMuPP[[1]], minWait, maxWait)
Wait(myJobPPMuPP[[1]], minWait, maxWait)
Wait(myJobCWPPIR[[1]], minWait, maxWait)
@

<<RetrieveJobs, results=HIDE, echo=FALSE>>=
RetrieveJob(myJobMuDP[[1]], c("phylip_interleaved.aln"))
RetrieveJob(myJobMuDF[[1]], c("fasta.aln"))
RetrieveJob(myJobMuDPP[[1]], c("phylip_pars.aln"))
RetrieveJob(myJobMuDPS[[1]], c("phylip_sequential.aln"))
RetrieveJob(myJobMuDC[[1]], c("clustalw.aln"))
RetrieveJob(myJobMuDM[[1]], c("msf.aln"))
RetrieveJob(myJobMuPP[[1]], c("phylip_interleaved.aln"))
RetrieveJob(myJobMuPF[[1]], c("fasta.aln"))
RetrieveJob(myJobMuPPP[[1]], c("phylip_pars.aln"))
RetrieveJob(myJobMuPPS[[1]], c("phylip_sequential.aln"))
RetrieveJob(myJobMuPC[[1]], c("clustalw.aln"))
RetrieveJob(myJobMuPM[[1]], c("msf.aln"))
RetrieveJob(myJobMaDF[[1]], c("mafft.fa"))
RetrieveJob(myJobMaDC[[1]], c("mafft.fa"))
RetrieveJob(myJobMaPF[[1]], c("mafft.fa"))
RetrieveJob(myJobMaPC[[1]], c("mafft.fa"))
RetrieveJob(myJobCWDP[[1]], c("clustalw2.fa"))
RetrieveJob(myJobCWDC[[1]], c("clustalw2.fa"))
RetrieveJob(myJobCWDN[[1]], c("clustalw2.fa"))
RetrieveJob(myJobCWDGCG[[1]], c("clustalw2.fa"))
RetrieveJob(myJobCWDGDE[[1]], c("clustalw2.fa"))
RetrieveJob(myJobCWDPIR[[1]], c("clustalw2.fa"))
RetrieveJob(myJobCWPP[[1]], c("clustalw2.fa"))
RetrieveJob(myJobCWPC[[1]], c("clustalw2.fa"))
RetrieveJob(myJobCWPN[[1]], c("clustalw2.fa"))
RetrieveJob(myJobCWPGCG[[1]], c("clustalw2.fa"))
RetrieveJob(myJobCWPGDE[[1]], c("clustalw2.fa"))
RetrieveJob(myJobCWPPIR[[1]], c("clustalw2.fa"))
RetrieveJob(myJobFaDMuP[[1]], c("fasttree.nwk"))
RetrieveJob(myJobFaDCWP[[1]], c("fasttree.nwk"))
RetrieveJob(myJobFaDMuF[[1]], c("fasttree.nwk"))
RetrieveJob(myJobFaDCWF[[1]], c("fasttree.nwk"))
RetrieveJob(myJobFaPMuP[[1]], c("fasttree.nwk"))
RetrieveJob(myJobFaPCWP[[1]], c("fasttree.nwk"))
RetrieveJob(myJobFaPMuF[[1]], c("fasttree.nwk"))
RetrieveJob(myJobFaPCWF[[1]], c("fasttree.nwk"))
RetrieveJob(myJobRDMuP[[1]], c("RAxML_bestTree.nwk"))
RetrieveJob(myJobRDCWP[[1]], c("RAxML_bestTree.nwk"))
RetrieveJob(myJobRPMuP[[1]], c("RAxML_bestTree.nwk"))
RetrieveJob(myJobRPCWP[[1]], c("RAxML_bestTree.nwk"))
RetrieveJob(myJobPDMuPP[[1]], c("outtree.nwk"))
RetrieveJob(myJobPPMuPP[[1]], c("outtree.nwk"))
@

\subsection{Genome Wide Association Study models}

{\fs
<<uploadfilegwas, echo=TRUE, eval=FALSE>>=
data(geno_test.tfam)
write(t(as.matrix(geno_test.tfam)), file="geno_test.tfam", ncolumns=length(geno_test.tfam), sep="\t")
UploadFile(local.file.name="geno_test.tfam")

data(geno_test.tped)
write(t(as.matrix(geno_test.tped)), file="geno_test.tped", ncolumns=length(geno_test.tped), sep="\t")
UploadFile(local.file.name="geno_test.tped")
@

<<uploadfilegwasres, echo=FALSE, eval=TRUE, results=verbatim>>=
data(geno_test.tfam)
write(t(as.matrix(geno_test.tfam)), file="geno_test.tfam", ncolumns=length(geno_test.tfam), sep="\t")
tryCatch(UploadFile(local.file.name="geno_test.tfam"), error=function(err){return(paste(err))})

data(geno_test.tped)
write(t(as.matrix(geno_test.tped)), file="geno_test.tped", ncolumns=length(geno_test.tped), sep="\t")
tryCatch(UploadFile(local.file.name="geno_test.tped"), error=function(err){return(paste(err))})
@
}

\subsection{PLINK Conversion}

\noindent {\ttfamily PLINKConversion(file.list="", file.path="", output.type="--recode", job.name=NULL, shared.username=NULL, print.curl=FALSE, version="plink-1.07u1", suppress.Warnings=FALSE)} \\

\noindent This function converts the standard PLINK file formats (Regular (ped/map), Transposed (tped/tfam), and Binary (bed/bim/fam)) to various other PLINK file formats.

{\fs
<<PLINKConversionR, echo=TRUE>>=
myJobPLINKCT <- PLINKConversion(file.list=list("geno_test.tfam","geno_test.tped"), job.name="PCT", 
                                out.basename="plinkout")
@

<<PLINKwait, echo=FALSE>>=
Wait(myJobPLINKCT[[1]], minWait, maxWait)
@

<<PLINKConversionT, echo=TRUE>>=
myJobPLINKCR <- PLINKConversion(file.list=list("plinkout.map","plinkout.ped"),
                                file.path=paste("analyses/", myJobPLINKCT[[2]], sep=""),
                                output.type="--recode --transpose", job.name="PCR")
@
}

\noindent There are many output files possible, \url{http://pngu.mgh.harvard.edu/~purcell/plink/reference.shtml#output}

\subsection{PLINK}

\noindent {\ttfamily PLINK(file.list="", file.path="", job.name=NULL, association.method="--assoc", no.sex=TRUE, args=NULL, print.curl=FALSE, multi.adjust=TRUE, version="plink-1.07u1", shared.username=NULL, suppress.Warnings=FALSE)} \\

\noindent {\ttfamily PLINK} is an open-source whole genome association analysis toolset, designed to perform a range of basic, large-scale analyses in a computationally efficient manner, check \url{http://pngu.mgh.harvard.edu/~purcell/plink/} for details.

{\fs
<<PLINK, echo=TRUE>>=
myJobPLINKT <- PLINK(file.list=list("geno_test.tfam","geno_test.tped"), job.name="PLINKT")
myJobPLINKR <- PLINK(file.list=list("plinkout.map","plinkout.ped"),
                     file.path=paste("analyses/", myJobPLINKCT[[2]], sep=""), job.name="PLINKR")
@
}

\noindent There are many output files possible, \url{http://pngu.mgh.harvard.edu/~purcell/plink/reference.shtml#output}



\subsection{FaST-LMM (Factored Spectrally Transformed Linear Mixed Models)}

\noindent {\ttfamily FaST\_LMM(input.file.list="", ALL.file.path="", print.curl=FALSE, sim.file.list=NULL, pheno.file.name=NULL, mpheno=1, args=NULL, covar.file.name=NULL, job.name=NULL, version="FaST-LMM-1.09u1", shared.username=NULL, suppress.Warnings=FALSE)} \\

\noindent {\ttfamily FaST-LMM (Factored Spectrally Transformed Linear Mixed Models)} is a program for performing genome-wide association studies (GWAS) on large data sets.  FaST-LMM is described more fully at \url{http://www.nature.com/nmeth/journal/v8/n10/abs/nmeth.1681.html}, and also at \url{http://fastlmm.codeplex.com/}

{\fs
<<FaST-LMM, echo=TRUE>>=
myJobFaST_LMMT <- FaST_LMM(input.file.list=list("geno_test.tfam","geno_test.tped"), 
                           job.name="FaST_LMMT")
myJobFaST_LMMR <- FaST_LMM(input.file.list=list("plinkout.map","plinkout.ped"),
                           ALL.file.path=paste("analyses/", myJobPLINKCT[[2]], sep=""), 
                           job.name="FaST_LMMR")
@
}

\noindent Not all information on the FaST-LMM model is here, see the FaST-LMM website \url{http://fastlmm.codeplex.com/}, or the FaST-LMM manual for more information.

\section{Creating workflows}

\noindent Finally, each of these steps can be combined to generate multi-step analyses. This has the benefit of reducing errors that can occur when manually running each application and, more importantly, ensures that results are reproducible. In the following example, a user starts with unaligned sequences on her or his local computer and ends with aligned sequences and a phylogenetic tree with all applications running on the iPlant servers. \\

\subsection{Workflow One}

\noindent This first workflow takes an amino acid fasta file, uses {\ttfamily MUSCLE} to get a {\ttfamily PHYLIP\_PARS} alignment type.  A couple things about this alignment; it is only available from {\ttfamily MUSCLE} and this alignment is very specific to the {\ttfamily PHYLIP 3.69} model.  The {\ttfamily PHYLIP} model then produces a tree.  Note: this is the only way to do this workflow.

{\fs
<<workflow1, echo=TRUE>>=
myJobW1MP <- Muscle("PROTEIN.fasta", aln.filetype="PHYLIP_PARS", job.name="muscleWORKFLOW1protein")
Wait(myJobW1MP[[1]], minWait, maxWait)
myJobW1PPP <- PHYLIP_Pars("phylip_pars.aln", file.path=paste("analyses/",myJobW1MP[[2]],
                          sep=""), type="PROTEIN", job.name="phylipWORKFLOW1protein")
Wait(myJobW1PPP[[1]], minWait, maxWait)
RetrieveJob(myJobW1PPP[[1]], c("outtree.nwk"))
read.tree(paste(getwd(), myJobW1PPP[[2]], "outtree.nwk", sep="/")) -> Tree
@
}

\begin{center}
\setkeys{Gin}{width=0.59\textwidth}
{\fs
<<PHYLIPtreePLOT, echo=FALSE, fig=TRUE, eps = FALSE, height=6, width=6>>=
plot(Tree)
@
}
\end{center}

\subsection{Workflow Two}

\noindent The second workflow takes the same amino acid fasta file and this time it uses {\ttfamily Mafft} to get a {\ttfamily FASTA} alignment type.  {\ttfamily MUSCLE} also can output a fasta alignment.  The {\ttfamily FastTree} model is then used to make the tree.

{\fs
<<workflow2, echo=TRUE>>=
myJobW2CWD <- Mafft("PROTEIN.fasta", type="PROTEIN", job.name="mafftPROTEINfasta")
Wait(myJobW2CWD[[1]], minWait, maxWait)
myJobW2FaD <- Fasttree("mafft.fa", type="PROTEIN", file.path=paste("analyses/",myJobW2CWD[[2]], 
                       sep=""), job.name="fasttreeCLUSTALWfasta")
Wait(myJobW2FaD[[1]], minWait, maxWait)
RetrieveJob(myJobW2FaD[[1]], c("fasttree.nwk"))
read.tree(paste(getwd(), myJobW2FaD[[2]], "fasttree.nwk", sep="/")) -> Tree
@
}

\begin{center}
\setkeys{Gin}{width=0.59\textwidth}
{\fs
<<FASTTREE1treePLOT, echo=FALSE, fig=TRUE, eps = FALSE, height=6, width=6>>=
plot(Tree)
@
}
\end{center}

\subsection{Workflow Three}

\noindent The third workflow is again dealing with {\ttfamily FastTree}.  {\ttfamily FastTree} can take either a FASTA alignment or a phylip interleaved alignment as inputs.  Now {\ttfamily ClustalW} takes a nucleotide fasta file to get a {\ttfamily PHYLIP INTERLEAVED} alignment type.  {\ttfamily MUSCLE} also can output that alignment.  The {\ttfamily FastTree} model is then used to make the tree.

{\fs
<<workflow3, echo=TRUE>>=
myJobW3MuP <- ClustalW("DNA.fasta", job.name="clustalwDNAfasta")
Wait(myJobW3MuP[[1]], minWait, maxWait)
myJobW3FaP <- Fasttree("clustalw2.fa", file.path=paste("analyses/",myJobW3MuP[[2]], sep=""), 
                        job.name="fasttreeMUSCLEdna")
Wait(myJobW3FaP[[1]], minWait, maxWait)
RetrieveJob(myJobW3FaP[[1]], c("fasttree.nwk"))
read.tree(paste(getwd(), myJobW3FaP[[2]], "fasttree.nwk", sep="/")) -> Tree
@
}

\begin{center}
\setkeys{Gin}{width=0.59\textwidth}
{\fs
<<FASTTREE2treePLOT, echo=FALSE, fig=TRUE, eps = FALSE, height=6, width=6>>=
plot(Tree)
@
}
\end{center}

\subsection{Workflow Four}

\noindent The fourth workflow is using {\ttfamily RAxML}.  {\ttfamily MUSCLE} takes a nucleotide fasta file to get a {\ttfamily PHYLIP INTERLEAVED} alignment type.  {\ttfamily ClustalW} also can output that alignment.  The {\ttfamily RAxML} model is then used to make the tree.

{\fs
<<workflow4, echo=TRUE>>=
myJobW4MuP <- Muscle("DNA.fasta", aln.filetype="PHYLIP_INT", job.name="muscleWORKFLOW4dna")
Wait(myJobW4MuP[[1]], minWait, maxWait)
myJobW4RP <- RAxML("phylip_interleaved.aln", file.path=paste("analyses/",myJobW4MuP[[2]], sep=""), 
                    job.name="raxmlWORKFLOW4dna")
Wait(myJobW4RP[[1]], minWait, maxWait)
RetrieveJob(myJobW4RP[[1]], c("RAxML_bestTree.nwk"))
read.tree(paste(getwd(), myJobW4RP[[2]], "RAxML_bestTree.nwk", sep="/")) -> Tree
@
}

\begin{center}
\setkeys{Gin}{width=0.59\textwidth}
{\fs
<<RAxMLtreePLOT, echo=FALSE, fig=TRUE, eps = FALSE, height=6, width=6>>=
plot(Tree)
@
}
\end{center}

<<last, echo=FALSE, eval=TRUE>>=
tryCatch(DeleteDir("hello"), error=function(err){return(paste(err))})
tryCatch(DeleteFile("DNA.fasta"), error=function(err){return(paste(err))})
tryCatch(DeleteFile("PROTEIN.fasta"), error=function(err){return(paste(err))})
tryCatch(DeleteFile("geno_test.tfam"), error=function(err){return(paste(err))})
tryCatch(DeleteFile("geno_test.tped"), error=function(err){return(paste(err))})
tryCatch(DeleteJob(ALL = TRUE), error=function(err){return(paste(err))})
@

\newpage


\bibliography{refs}
\bibliographystyle{jasa}

\end{document}



